go/backend/stock/memory/memory.go

```diff
@@ -251,13 +282,83 @@ func (s *inMemoryStock[I, V]) Close() error {
 	return s.Flush()
 }
 
+func (s *inMemoryStock[I, V]) GuaranteeCheckpoint(checkpoint checkpoint.Checkpoint) error {
+	// If requested checkpoint is the stock's current checkpoint, everything is fine.
+	if s.checkpoint == checkpoint {
+		return nil
+	}
+
+	// If the requested checkpoint is pending, it needs to be completed.
+	if s.checkpoint+1 == checkpoint {
+		return s.Commit(checkpoint)
+	}
+
+	// Otherwise, the stock is in an inconsistent state.
+	return fmt.Errorf("unable to guarantee availability of checkpoint %d", checkpoint)
+}
+
+func (s *inMemoryStock[I, V]) Prepare(commit checkpoint.Checkpoint) error {
+	if s.checkpoint+1 != commit {
+		return fmt.Errorf("unable to prepare checkpoint %d, expected %d", commit, s.checkpoint+1)
+	}
+	if err := s.Flush(); err != nil {
+		return err
+	}
+	return utils.WriteJsonFile(filepath.Join(s.directory, "prepare.json"), checkpointData{
+		Checkpoint:     commit,
+		NumValues:      len(s.values),
+		FreeListLength: len(s.freeList),
+	})
+}
+
+func (s *inMemoryStock[I, V]) Commit(checkpoint checkpoint.Checkpoint) error {
+	prepareFile := filepath.Join(s.directory, "prepare.json")
+	commitFile := filepath.Join(s.directory, "commit.json")
+	meta, err := utils.ReadJsonFile[checkpointData](prepareFile)
+	if err != nil {
+		return err
+	}
+	if meta.Checkpoint != checkpoint {
+		return fmt.Errorf("unable to commit checkpoint %d, expected %d", checkpoint, meta.Checkpoint)
+	}
+	s.checkpoint = checkpoint
+	s.checkpointValueListLength = I(meta.NumValues)
+	s.checkpointFreeListLength = meta.FreeListLength
+	return os.Rename(prepareFile, commitFile)
+}
+
+func (s *inMemoryStock[I, V]) Abort(commit checkpoint.Checkpoint) error {
+	if s.checkpoint+1 != commit {
+		return fmt.Errorf("unable to abort checkpoint %d, expected %d", commit, s.checkpoint+1)
+	}
+	s.checkpointValueListLength = s.backupCheckpointValueListLength
+	s.checkpointFreeListLength = s.backupCheckpointFreeListLength
+	prepareFile := filepath.Join(s.directory, "prepare.json")
+	return os.Remove(prepareFile)
+}
```

Comment:
Lets keep the code consistent and also create constants for `prepare.json` and `commit.json`